hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/AggregateAbstractPhysicalRule.java

```diff
@@ -70,71 +76,489 @@ protected AggregateAbstractPhysicalRule(Config config) {
-                    Object[] row = new Object[aggregations.size()];
-                    for (int i = 0; i < aggregations.size(); i++) {
-                        row[i] = aggregations.get(i).collect();
-                    }
-                    return new JetSqlRow(Contexts.getCastedThreadContext().serializationService(), row);
-                });
+    public static class AggregateAvgSupplier implements IdentifiedDataSerializable,
+            SupplierEx<SqlAggregation> {
+        private boolean distinct;
+        private QueryDataType avgOperandType;
+
+        public AggregateAvgSupplier() {
+        }
+
+        public AggregateAvgSupplier(boolean distinct, QueryDataType avgOperandType) {
+            this.distinct = distinct;
+            this.avgOperandType = avgOperandType;
+        }
+
+        @Override
+        public SqlAggregation getEx() throws Exception {
+            return AvgSqlAggregations.from(avgOperandType, distinct);
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeBoolean(distinct);
+            out.writeObject(avgOperandType);
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+            distinct = in.readBoolean();
+            avgOperandType = in.readObject();
+        }
+
+        @Override
+        public int getFactoryId() {
+            return JetSqlSerializerHook.F_ID;
+        }
+
+        @Override
+        public int getClassId() {
+            return JetSqlSerializerHook.AGGREGATE_AVG_SUPPLIER;
+        }
+    }
+
+    public static class AggregateSumSupplier implements IdentifiedDataSerializable,
+            SupplierEx<SqlAggregation> {
+        private boolean distinct;
+        private QueryDataType sumOperandType;
+
+        public AggregateSumSupplier() {
+        }
+
+        public AggregateSumSupplier(boolean distinct, QueryDataType sumOperandType) {
+            this.distinct = distinct;
+            this.sumOperandType = sumOperandType;
+        }
+
+        @Override
+        public SqlAggregation getEx() throws Exception {
+            return SumSqlAggregations.from(sumOperandType, distinct);
+        }
+
+        @Override
+        public void writeData(ObjectDataOutput out) throws IOException {
+            out.writeBoolean(distinct);
+            out.writeObject(sumOperandType);
+        }
+
+        @Override
+        public void readData(ObjectDataInput in) throws IOException {
+           distinct = in.readBoolean();
+           sumOperandType = in.readObject();
+        }
+
+        @Override
+        public int getFactoryId() {
+            return JetSqlSerializerHook.F_ID;
+        }
+
+        @Override
+        public int getClassId() {
+            return JetSqlSerializerHook.AGGREGATE_SUM_SUPPLIER;
+        }
+    }
+
+    public static final class AggregateCountTFSupplier implements IdentifiedDataSerializable,
```

Comment:
I'd unify the `AggregateCount*Supplier` classes and rather put two boolean fields into the class and pass those to `CountSqlAggregations.from`.

Reply:
I cna do this, but then I cannot do singleton for each class. Now I have three instancess on the heap, adding fields wil require creating new objects during code execution.

Reply:
Using singletons in initialization isn't that big deal. We don't created even hundreds of instances of them. I prefer readability. Even the name `TFSupplier` is ... mysterious?